apiserver/facades/client/charms/client.go

```diff
@@ -288,32 +290,68 @@ func (a *API) AddCharmWithAuthorization(args params.AddCharmWithAuth) (params.Ch
+// compatibilityApplicationDeployArgs ensures that Deploy calls from
+// clients such as juju 3.x, python-libjuju 3.x, etc. work against a 2.9.x
+// controller. To support these clients, we make no assumptions about
+// which of the 3 potential places the OS is specified. We check all 3,
+// assert any provded match, and fill in the remain series.
+//
+// In juju 3.x, params.AddCharmWithAuth was changed to remove series
 // however, the facade version was not changed, nor was the name of
 // the params.AddCharmWithAuth changed. Thus it appears you can use
 // a juju 3.x client to deploy from a juju 2.9 controller, however it
-// fails because the series was not found. Make those corrections here.
+// fails because the series was not found.
+//
+// NOTE: We do not mess with charm url because this is dangerous. Mutating
+// charm url can lead to unexpected results
 func compatibilityAddCharmWithAuthArg(arg params.AddCharmWithAuth) (params.AddCharmWithAuth, error) {
-	origin := arg.Origin
-	if origin.Series != "" {
-		return arg, nil
+	s, err := getUnifiedSeries(arg)
+	if err != nil {
+		return arg, errors.Annotatef(err, "cannot add charm")
 	}
-	originSeries, err := series.GetSeriesFromChannel(origin.Base.Name, origin.Base.Channel)
+	b, err := series.GetBaseFromSeries(s)
 	if err != nil {
-		return arg, err
+		return arg, errors.Trace(err)
 	}
-	arg.Origin.Series = originSeries
-	arg.Series = originSeries
+
+	arg.Series = s
+	arg.Origin.Series = s
+	arg.Origin.Base.Name = b.Name
+	arg.Origin.Base.Channel = b.Channel.String()
+
 	return arg, nil
 }
 
+// getUnifiedSeries checks the multiple places a series can be provided in a params.AddCharmWithAuth,
+// checks they match if they're provided, and returns the matching series
+func getUnifiedSeries(arg params.AddCharmWithAuth) (string, error) {
+	argSeries := arg.Series
+	originSeries := arg.Origin.Series
+	originBaseSeries, _ := series.GetSeriesFromChannel(arg.Origin.Base.Name, arg.Origin.Base.Channel)
+
+	s := argSeries
+	if s == "" {
+		s = originSeries
+	}
+	if s == "" {
+		s = originBaseSeries
+	}
+	if s == "" {
+		return "", errors.Errorf("no series provided")
+	}
+	if (argSeries != "" && s != argSeries) || (originSeries != "" && s != originSeries) || (originBaseSeries != "" && s != originBaseSeries) {
+		seriesString := strings.Join([]string{argSeries, originSeries, originBaseSeries}, " ")
+		seriesString = strings.TrimSpace(strings.ReplaceAll(seriesString, "  ", " "))
+		return "", errors.Errorf("provided series [%s] do not match", seriesString)
+	}
+	return s, nil
+}
+
 func (a *API) addCharmWithAuthorization(args params.AddCharmWithAuth) (params.CharmOriginResult, error) {
 	var err error
 	args, err = compatibilityAddCharmWithAuthArg(args)
 	if err != nil {
-		return params.CharmOriginResult{}, errors.Annotatef(err, "compatibility updates of origin failed")
+		return params.CharmOriginResult{}, errors.Annotatef(err, "compatibility updates failed")
```

Comment:
Agreed. I don't think there's much the `addCharm` layer can add to the error here, especially since I've done work to make the errors in `getUnifiedSeries` more readable. As such, I've replaced with `errors.Trace`